securedrop_client/api_jobs/downloads.py

```diff
@@ -2,171 +2,111 @@
-        db_object = session.query(self.submission_type).filter_by(uuid=self.submission_uuid).one()
+        db_object = self.get_db_object(session)
 
         if db_object.is_decrypted:
             return db_object.uuid
 
         if db_object.is_downloaded:
-            self._decrypt_file(session, db_object, os.path.join(self.data_dir, db_object.filename))
+            self._decrypt(os.path.join(self.data_dir, db_object.filename), db_object, session)
             return db_object.uuid
 
         self._download(api_client, db_object, session)
-        self._decrypt_file(session, db_object, os.path.join(self.data_dir, db_object.filename))
+        self._decrypt(os.path.join(self.data_dir, db_object.filename), db_object, session)
         return db_object.uuid
 
-    def _download(self, api_client: API, db_object: File, session: Session) -> None:
+    def _download(self, api_client: API, db_object: Union[File, Message], session: Session) -> None:
         '''
-        Download the file. Check file integrity and move it to the data directory before marking it
-        as downloaded.
+        Download the encrypted file. Check file integrity and move it to the data directory before
+        marking it as downloaded.
 
         Note: On Qubes OS, files are downloaded to ~/QubesIncoming.
         '''
-        etag, download_path = self._make_call(db_object, api_client)
-
-        if not self._check_file_integrity(etag, download_path):
-            raise RuntimeError('Downloaded file had an invalid checksum.')
+        try:
+            etag, download_path = self.call_download_api(api_client)
 
-        data_path = os.path.join(self.data_dir, db_object.filename)
-        shutil.move(download_path, data_path)
+            if not self._check_file_integrity(etag, download_path):
+                raise RuntimeError('Downloaded file had an invalid checksum.')
 
-        mark_as_downloaded(type(db_object), db_object.uuid, session)
+            shutil.move(download_path, os.path.join(self.data_dir, db_object.filename))
+            mark_as_downloaded(type(db_object), db_object.uuid, session)
+            logger.info("File downloaded: {}".format(db_object.filename))
+        except BaseError as e:
+            logger.debug("Failed to download file: {}".format(db_object.filename))
+            raise e
 
-    def _make_call(self, db_object: Union[File, Message], api_client: API) -> Tuple[str, str]:
-        sdk_obj = sdclientapi.Submission(uuid=db_object.uuid)
-        sdk_obj.filename = db_object.filename
-        sdk_obj.source_uuid = db_object.source.uuid
-        return api_client.download_submission(sdk_obj)
+    def _decrypt(self, filepath: str, db_object: Union[File, Message], session: Session) -> None:
+        '''
+        Decrypt the file located at the given filepath and mark it as decrypted.
+        '''
+        try:
+            self.call_decrypt(filepath, session)
+            mark_as_decrypted(type(db_object), db_object.uuid, session)
```

Comment:
I like your `set_message_or_reply_content` suggestion. It makes it clear that we don't set content for Files and it separates marking the field as decrypted in case there are further checks we want to do beforehand. For instance, with file downloads we check the file integrity and move it before marking it as downloaded. 

Unfortunately there is this database constraint:
```py
        CheckConstraint('CASE WHEN is_decrypted = 0 THEN content IS NULL ELSE 1 END',
                        name='messages_compare_is_decrypted_vs_content'),
```

So we can't separate setting content from marking something as decrypted until this constraint is removed. I think the code is a lot clear if we do remove this constraint and use this `mark_as_downloaded` `mark_as_decrypted` consistency.

Otherwise the code becomes less consistent and less modular (instead of just decrypting a file and storing the content on the filesystem or in the db, `call_decrypt` will do this _and_ mark the db object as decrypted).

But it's understandable if we need to create a work-around for now and followup on cleaning this up later.

Reply:
oh yeah good point - OK, I'm for ditching this database constraint for the sake of modularity